### PR TITLE
Lvnod-119_Create Сomment component

### PR DESCRIPTION
**Pull request by shyyka**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/LVNOD-XXX)
* [JIRA subtask](https://ssu-jira.softserveinc.com/browse/LVNOD-XXX)

## Code reviewers

- [x] @yarvanya
- [x] @denisbodnar
- [x] @yuliannakorol

### Second Level Review

- [ ] @lhalam
- [ ] @4asof4ik

### CODE REVIEW

File: app/src/containers/Comments/CommentForm.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
+import { FormGroup, FormControl, Button, Form } from 'react-bootstrap';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import * as createCommentActions from '../../redux/commentCreateReducers';
+import messages from '../../helper/messages';
+import getAuthor from './../../helper/getAuthor';
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentForm.js]: you import helper in 6 line

File: app/src/containers/Comments/CommentForm.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import './Comments.scss';
+
+export class CommentForm extends React.Component{
+
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentForm.js]: drop line

File: app/src/containers/Comments/CommentForm.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  constructor(props){
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentForm.js]: space before {

File: app/src/containers/Comments/CommentForm.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  formSubmit = e => {
+    e.preventDefault();
+
+    const commentData = {
+      body: this.state.body,
+      parent_id: this.props.parent_id
+    };
+    const { eventId, giftId} = this.props;
+    this.props.actions.createComment(eventId, giftId, commentData);
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentForm.js]: 30-32 add into 29 line

File: app/src/containers/Comments/CommentForm.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    this.props.getComments();
+    this.props.hideForm && this.props.hideForm() || null;
+
+    this.setState({
+      body: '',
+      parent_id: null
+    });
+  }
+
+  render(){
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentForm.js]: space {

File: app/src/containers/Comments/CommentForm.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    const error = this.state.body && !this.inputValid() && messages.maxTextLength || '';
+    const { firstName, lastName } = this.props.author;
+
+    return (
+      <div className="comment-wrapper">
+        <div className = "comment-head">{`${firstName} ${lastName}`}</div>
+        <div className="content">
+          <img className="avatar" src={ this.props.author.avatar } />
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentForm.js]: select one code style {} with space inside or not

File: app/src/containers/Comments/CommentItem.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import { Button, Glyphicon } from 'react-bootstrap';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import * as deleteCommentActions from '../../redux/commentDeleteReducers';
+import CommentForm from './CommentForm';
+import getAuthor from './../../helper/getAuthor';
+import './Comments.scss';
+
+export class CommentItem extends React.Component{
+  constructor(props){
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentItem.js]: check this file

File: app/src/containers/Comments/CommentItem.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+              <div className="reply">
+                { this.props.user }<p> { commentsDate } | </p>
+                <a onClick={()=>this.setState({ showForm: !this.state.showForm })}>
+                  <Glyphicon glyph="share-alt" /> Reply </a>
+                {
+                  this.state.showForm ?
+                    <CommentForm
+                      user={ getAuthor(this.props.comment) }
+                      parent_id={ this.props.comment.id }
+                      author={ this.props.author }
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentItem.js]: write const {} = this.props

File: app/src/helper/getAuthor.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+const getAuthor = comment => {
+  const { first_name: firstName, last_name: lastName } = comment.User.Profile;
+  return firstName + ' ' + lastName;
```
Comment (4asof4ik) [inline app/src/helper/getAuthor.js]: use `${}`

File: app/src/containers/Event/EventDetails.js [hunk 10]
```diff
@@ -84,9 +85,14 @@ class EventDetails extends React.Component {
...
-              <Link className='list-group-item' to={'/events/' + id + '/gifts'}>
-                  Gifts
+              <div>
+                <PageHeader className='text-center'> Gifts </PageHeader>
+              </div>
+              <Link className='list-group-item' to={`/events/${id}/gifts`}>
+                  Sample Gifts
               </Link>
+              { this.props.children }
+              <Comments eventId={id} giftId={1} author={this.props.owner} />
```
Comment (4asof4ik) [inline app/src/containers/Event/EventDetails.js]: gift id hardcode

File: app/src/containers/Comments/CommentItem.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            <Glyphicon glyph="remove" /></a>
+          <div>
+            <div className = "comment-head">
+              { getAuthor(this.props.comment) }
+            </div>
+            <div className="content">
+              <img className="avatar" src="{ this.props.author.avatar }"  />
+              <div className="message">{ this.props.comment.body }</div>
+              <div className="reply">
+                { this.props.user }<p> { commentsDate } | </p>
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentItem.js]: separate

File: app/src/containers/Comments/CommentItem.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                <a onClick={()=>this.setState({ showForm: !this.state.showForm })}>
+                  <Glyphicon glyph="share-alt" /> Reply </a>
+                {
+                  this.state.showForm ?
+                    <CommentForm
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentItem.js]: create const and use them

File: app/src/containers/Comments/CommentItem.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+                      giftId={ this.props.giftId }
+                      getComments={ this.props.getComments }
+                      hideForm={ () => this.setState({ showForm: false }) }/> :
+                    null
+                }
+              </div>
+            </div>
+          </div>
+          {
+            this.props.comment.children ? (
```
Comment (4asof4ik) [inline app/src/containers/Comments/CommentItem.js]: create function which will be return this

File: app/src/redux/commentCreateReducers.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
+const COMMENT_CREATED = 'COMMENT_CREATED';
+const CREATE_COMMENT_FAILURE = 'CREATE_COMMENT_FAILURE';
+import  { API }  from './../helper/constants';
```
Comment (4asof4ik) [inline app/src/redux/commentCreateReducers.js]: one time helper import  chack all files

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/containers/Comments/CommentItem.js [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  delComment = () => {
+    const { eventId, giftId, comment } = this.props;
+    const comment_id = comment.id;
+    this.props.actions.deleteComment(eventId, giftId, comment_id);
+    this.props.getComments();
+  };
+
+  render(){
```
Comment (yarvanya) [inline app/src/containers/Comments/CommentItem.js]: Space before {

File: app/src/containers/Comments/CommentItem.js [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+                {
+                  this.state.showForm ?
+                    <CommentForm
+                      user={ getAuthor(this.props.comment) }
+                      parent_id={ this.props.comment.id }
+                      author={ this.props.author }
+                      eventId={ this.props.eventId }
+                      giftId={ this.props.giftId }
+                      getComments={ this.props.getComments }
+                      hideForm={ () => this.setState({ showForm: false }) }/> :
```
Comment (yarvanya) [inline app/src/containers/Comments/CommentItem.js]: Space before  />

File: app/src/containers/Comments/Comments.js [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    this.getComments();
+  }
+
+  getComments = () => {
+    const { eventId, giftId } = this.props;
+    this.props.actions.retrieveComments(eventId, giftId);
+  };
+
+  render() {
+
```
Comment (yarvanya) [inline app/src/containers/Comments/Comments.js]: Delete this empty blank.

File: app/src/containers/Comments/Comments.scss [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+}
+img.avatar {
+  margin: 0 5px;
+  width: 55px;
+  height: 55px;
+  vertical-align: top;
+}
+
+.reply {
+  border-top: 1px solid #ddedeb;
```
Comment (yarvanya) [inline app/src/containers/Comments/Comments.scss]: Add this color to the constants in common.scss .

File: app/src/containers/Comments/Comments.scss [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  display: flex;
+  background-color: #f3fffd;
```
Comment (yarvanya) [inline app/src/containers/Comments/Comments.scss]: And this.

File: app/src/redux/commentDeleteReducers.js [hunk 20]
```diff
@@ -0,0 +1,46 @@
+const COMMENT_DELETED = 'COMMENT_DELETED';
```
Comment (yarvanya) [inline app/src/redux/commentDeleteReducers.js]: First should be importing and after that declareting constants.

File: app/src/redux/commentCreateReducers.js [hunk 21]
```diff
@@ -0,0 +1,47 @@
+const COMMENT_CREATED = 'COMMENT_CREATED';
```
Comment (yarvanya) [inline app/src/redux/commentCreateReducers.js]: First should be importing and after that declareting constants.

File: app/src/redux/commentsFetchReducers.js [hunk 22]
```diff
@@ -0,0 +1,51 @@
+const RETRIEVE_COMMENTS = 'RETRIEVE_COMMENTS';
```
Comment (yarvanya) [inline app/src/redux/commentsFetchReducers.js]: First should be importing and after that declareting constants.

Comment (shyyka) [reply app/src/redux/commentsFetchReducers.js]: Ok

File: app/src/redux/commentCreateReducers.js (hunk 21 above)
Comment (shyyka) [reply app/src/redux/commentCreateReducers.js]: Done

File: app/src/containers/Event/EventDetails.js (hunk 10 above)
Comment (shyyka) [reply app/src/containers/Event/EventDetails.js]: Temporary, Comments are attached to the test "Gift" component so far which will be replaced by adding a working one

File: app/src/containers/Comments/CommentForm.js (hunk 1 above)
Comment (shyyka) [reply app/src/containers/Comments/CommentForm.js]: Done

File: app/src/containers/Comments/CommentForm.js (hunk 4 above)
Comment (shyyka) [reply app/src/containers/Comments/CommentForm.js]: Done

File: app/src/containers/Comments/Comments.js (hunk 17 above)
Comment (shyyka) [reply app/src/containers/Comments/Comments.js]: We talked on Friday that we left the blank line between render & return

File: app/src/containers/Comments/CommentForm.js (hunk 2 above)
Comment (shyyka) [reply app/src/containers/Comments/CommentForm.js]: Done

File: app/src/containers/Comments/CommentForm.js (hunk 6 above)
Comment (shyyka) [reply app/src/containers/Comments/CommentForm.js]: Ok

File: app/src/containers/Comments/CommentItem.js (hunk 7 above)
Comment (shyyka) [reply app/src/containers/Comments/CommentItem.js]: Done

File: app/src/containers/Comments/CommentItem.js (hunk 12 above)
Comment (shyyka) [reply app/src/containers/Comments/CommentItem.js]: Done

File: app/src/redux/commentDeleteReducers.js [hunk 23]
```diff
@@ -0,0 +1,46 @@
+import  { API, request }  from './../helper';
+
+const COMMENT_DELETED = 'COMMENT_DELETED';
+const DELETE_COMMENT_FAILURE = 'DELETE_COMMENT_FAILURE';
+
+export const deleteComment = (eventId, giftId, comment_id) => {
+  return dispatch => {
+    request()
+      .delete(API.URL + `/api/event/${eventId}/gift/${giftId}/comment/${comment_id}`)
```
Comment (4asof4ik) [inline app/src/redux/commentDeleteReducers.js]: use for all (include api.url)

File: app/src/redux/commentDeleteReducers.js [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+      .end((err, res) => {
+        if (err) {
```
Comment (4asof4ik) [inline app/src/redux/commentDeleteReducers.js]: improve check all reducers

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

File: app/src/redux/commentDeleteReducers.js (hunk 24 above)
Comment (shyyka) [reply app/src/redux/commentDeleteReducers.js]: Done

File: app/src/redux/commentDeleteReducers.js (hunk 23 above)
Comment (shyyka) [reply app/src/redux/commentDeleteReducers.js]: Done

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

File: app/src/containers/Event/EventDetails.js [hunk 25]
```diff
@@ -85,7 +85,7 @@ class EventDetails extends React.Component {
             </Tab>
             <Tab eventKey={3} title='Gifts'>
               <h2>Gift list</h2>
-              <Gift id={id}/>
+              <Gift id={id} author={this.props.owner} />
```
Comment (denisbodnar) [inline app/src/containers/Event/EventDetails.js]: shorten this.props

File: app/src/containers/Gift/Gift.js [hunk 26]
```diff
@@ -15,7 +15,7 @@ class Gift extends React.Component {
     return (
       <div>
         <AddGift id={id} actions={actions} gifts={gifts}/>
-        <GiftList id={id} actions={actions} gifts={gifts}/>
+        <GiftList id={id} author={this.props.author} actions={actions} gifts={gifts}/>
```
Comment (denisbodnar) [inline app/src/containers/Gift/Gift.js]: same

**Review by denisbodnar — DISMISSED** [review]

File: app/src/redux/reducer.js [hunk 27]
```diff
@@ -10,6 +10,9 @@ import newEvent from './newEventReducers';
 import editEvent from './editEventReducers';
 import contactInfo from './contactInfo';
 import gift from './gift';
+import commentsList from './commentsFetchReducers';
```
Comment (4asof4ik) [inline app/src/redux/reducer.js]: create one reducer and check how it should looks in Denys PR (action and reducer is separated )

File: app/src/redux/commentsFetchReducers.js [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+
+const RETRIEVE_COMMENTS = 'RETRIEVE_COMMENTS';
+const RETRIEVE_COMMENTS_SUCCESS = 'RETRIEVE_COMMENTS_SUCCESS';
+const RETRIEVE_COMMENTS_FAILURE = 'RETRIEVE_COMMENTS_FAILURE';
+
+export const retrieveComments = (eventId, giftId) => dispatch =>
+  request()
+    .get(`${API.URL}/api/event/${eventId}/gift/${giftId}/comments`)
+    .end((err, res) => err &&
+        dispatch({
```
Comment (4asof4ik) [inline app/src/redux/commentsFetchReducers.js]: create const for action and use them here check all PR

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/redux/reducer.js (hunk 27 above)
Comment (shyyka) [reply app/src/redux/reducer.js]: Done

File: app/src/redux/commentsFetchReducers.js (hunk 28 above)
Comment (shyyka) [reply app/src/redux/commentsFetchReducers.js]: Done

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by denisbodnar — APPROVED** [review]

**Review by yarvanya — APPROVED** [review]

**Review by yuliannakorol — APPROVED** [review]

**Pull request merged**